subsys/net/lib/sockets/sockets.c

```diff
@@ -1522,6 +1522,20 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
 
 				return 0;
 			}
+			break;
+
+		case SO_PROTOCOL: {
+			int proto = (int)net_context_get_ip_proto(ctx);
+
+			if (*optlen >= sizeof(int)) {
+				*(int *)optval = proto;
+				*optlen = sizeof(int);
+			} else {
+				memcpy(optval, &proto, *optlen);
+			}
```

Comment:
Same comment here, this looks unnecessary complex and error prone if user supplies anything else than int as optval.